src/tools/clippy/clippy_utils/src/consts.rs

```diff
@@ -616,10 +617,10 @@ pub fn miri_to_const(result: ty::Const<'_>) -> Option<Constant> {
             },
             _ => None,
         },
-        ty::ConstKind::Value(ConstValue::ByRef { alloc, offset: _ }) => match result.ty().kind() {
+        mir::ConstantKind::Val(ConstValue::ByRef { alloc, offset: _ }, _) => match result.ty().kind() {
             ty::Array(sub_type, len) => match sub_type.kind() {
-                ty::Float(FloatTy::F32) => match miri_to_const(*len) {
-                    Some(Constant::Int(len)) => alloc
+                ty::Float(FloatTy::F32) => match len.try_eval_usize(tcx, ty::ParamEnv::empty()) {
```

Comment:
you don't have the right `param_env` for eval.

`len.val().try_to_machine_usize(tcx)` does not try to evaluate it if it's `Unevaluated`.

Reply:
so this would (probably) ICE for the following cause we try to compute `std::mem::size_of::<T>()` for a `T` which might not be sized (as it isn't mentioned in the `param_env`).
```rust
struct Foo<T>(T);

impl<T> Foo<T> {
    const ASSOC: [u8; std::mem::size_of::<T>()] = [0; std::mem::size_of::<T>()];
}

fn foo<T>() {
    Foo::<T>::ASSOC
}
```

Reply:
The constant is guaranteed to already be evaluated here, is it not? So we could just get the valtree and convert that to usize?

Reply:
well, that's what `len.val().try_to_machine_usize(tcx)` is doing :laughing: except that is has checks that the value of the `ScalarInt` fits into a `usize` which is redundant as it's an invariant of `ScalarInt`

Reply:
Sorry, can't think anymore today, have to do that tomorrow.